pyemma/coordinates/transform/transformer.py

```diff
@@ -35,15 +35,89 @@
 __all__ = ['Transformer']
 __author__ = 'noe, marscher'
 
+class TransformerIteratorContext(object):
+
+    def __init__(self, stride=1, lag=0):
+        self._lag = lag
+        self.__init_stride(stride)
+
+    def __init_stride(self, stride):
+        self._stride = stride
+        if isinstance(stride, np.ndarray):
+            keys = stride[:, 0]
+            self._trajectory_keys, self._trajectory_lengths = np.unique(keys, return_counts=True)
+        else:
+            self._trajectory_keys = None
+        self._uniform_stride = TransformerIteratorContext.is_uniform_stride(stride)
+        if not self.uniform_stride and not self.is_stride_sorted():
+            raise ValueError("Currently only sorted arrays allowed for random access")
+
+    def ra_indices_for_traj(self, traj):
+        """
+        Gives the indices for a trajectory file (without changing the order within the trajectory itself).
+        :param traj:
+        :return:
+        """
+        assert not self.uniform_stride, "requested random access indices, but is in uniform stride mode"
+        return self._stride[self._stride[:, 0] == traj][:, 1] if traj in self.traj_keys else np.array([])
```

Comment:
I think the only way to speed things up is to stream through the file once (and if order needs to be preserved, have a mapping or something) and collect the data with seek and whence. In my code it is like this (leaving out chunk sizes):
```python
# some initialization
x_prev = 0
curr_size = 0
# this loop groups together subsequent indices, so [1,2,3,5,6,8,9] becomes [1,2,3],[5,6],[8,9]
for k, g in groupby(enumerate(stride), lambda (a, b): a-b):
    # this here is the current group, i.e., [5,6]
    grouped_stride = map(itemgetter(1), g)
    # some offset because the trajectory begin has some special handling
    seek_offset = (1 if x_prev != 0 else 0)
    # this is how far we need to seek
    seek_to = grouped_stride[0] - x_prev - seek_offset
    # do the seek
    f.seek(seek_to, whence=1)
    # remember to which index we read the trajectory so we can calculate seek_to for the next group
    x_prev = grouped_stride[-1]
    # how many frames do we read
    group_size = len(grouped_stride)
    # read it!
    traj = _get_local_traj_object(atom_indices, extension, f, group_size, topology, **kwargs)
```
after having read the trajectories for the index groups, they get concatenated and yielded.